File: src/shared/contracts/operation-progress.ts
export type MergeProgressPhase =
    | 'preparing-documents'
    | 'merging-pages'
    | 'optimizing-images'
    | 'saving';

export type PageCompositionProgressPhase = 'validating' | 'composing' | 'saving';

export const OPERATION_PROGRESS_EVENT = 'operation-progress';
export const OPERATION_PROGRESS_VERSION = 1 as const;

/** Versioned progress contract shared by workflow operation transports. */
export type OperationProgressPayload = {
    version: typeof OPERATION_PROGRESS_VERSION;
    operation: 'merge' | 'page-composition';
    phase: string;
    percentage: number;
};

export function isOperationProgressPayload(value: unknown): value is OperationProgressPayload {
    if (typeof value !== 'object' || value === null) return false;

    const payload = value as Record<string, unknown>;
    return (
        payload.version === OPERATION_PROGRESS_VERSION &&
        (payload.operation === 'merge' || payload.operation === 'page-composition') &&
        typeof payload.phase === 'string' &&
        typeof payload.percentage === 'number' &&
        Number.isInteger(payload.percentage) &&
        payload.percentage >= 0 &&
        payload.percentage <= 100
    );
}

export type MergeOperationProgressPayload = OperationProgressPayload & {
    operation: 'merge';
    phase: MergeProgressPhase;
};

export function isMergeOperationProgressPayload(
    value: unknown,
): value is MergeOperationProgressPayload {
    if (!isOperationProgressPayload(value) || value.operation !== 'merge') return false;

    return (
        value.phase === 'preparing-documents' ||
        value.phase === 'merging-pages' ||
        value.phase === 'optimizing-images' ||
        value.phase === 'saving'
    );
}

export type PageCompositionOperationProgressPayload = OperationProgressPayload & {
    operation: 'page-composition';
    phase: PageCompositionProgressPhase;
};

export function isPageCompositionOperationProgressPayload(
    value: unknown,
): value is PageCompositionOperationProgressPayload {
    if (!isOperationProgressPayload(value) || value.operation !== 'page-composition') return false;
    return value.phase === 'validating' || value.phase === 'composing' || value.phase === 'saving';
}

File: src-tauri/src/shared/operation_progress.rs
use serde::Serialize;

/// Versioned progress payload shared by workflow-owned operation transports.
///
/// The event name is stable across workflows; `operation` and `phase` keep the payload
/// extensible without making one workflow depend on another workflow's phase vocabulary.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct OperationProgressEnvelope {
    pub(crate) version: u8,
    pub(crate) operation: &'static str,
    pub(crate) phase: &'static str,
    pub(crate) percentage: u8,
}

pub(crate) const OPERATION_PROGRESS_EVENT: &str = "operation-progress";

impl OperationProgressEnvelope {
    pub(crate) const VERSION: u8 = 1;

    pub(crate) const fn merge(phase: &'static str, percentage: u8) -> Self {
        Self {
            version: Self::VERSION,
            operation: "merge",
            phase,
            percentage,
        }
    }

    pub(crate) const fn page_composition(phase: &'static str, percentage: u8) -> Self {
        Self {
            version: Self::VERSION,
            operation: "page-composition",
            phase,
            percentage,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::OperationProgressEnvelope;

    #[test]
    fn serializes_the_versioned_merge_envelope() {
        let payload = OperationProgressEnvelope::merge("merging-pages", 42);

        assert_eq!(
            serde_json::to_value(payload).expect("progress envelope should serialize"),
            serde_json::json!({
                "version": 1,
                "operation": "merge",
                "phase": "merging-pages",
                "percentage": 42,
            })
        );
    }

    #[test]
    fn serializes_page_composition_without_merge_phases() {
        let payload = OperationProgressEnvelope::page_composition("composing", 55);
        assert_eq!(
            serde_json::to_value(payload).unwrap(),
            serde_json::json!({
                "version": 1,
                "operation": "page-composition",
                "phase": "composing",
                "percentage": 55,
            })
        );
    }
}
